Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,21 +21,58 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return totalAmpHoursUsed_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    if (current_ < 0)
+    {
+       return true;
+    }
+    else
+    {
+        return false;
+    }
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
-}
+    int ms_remaining;
+    if (isCharging())
+    {
+      ms_remaining = totalAmpHoursUsed() / averageCurrent_ * HOURS_TO_MS_CONVERSION;
+    }
+    else
+    {
+      ms_remaining = (BATTERY_AMP_HOUR_CAPACITY - totalAmpHoursUsed()) / averageCurrent_ * HOURS_TO_MS_CONVERSION;
+    }
+
+    return QTime(0, 0, 0, 0).addMSecs(ms_remaining);
 
+}
+/*running average needs to be just the one after it, so I need to write the average as the current + past current divided by 2*/
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // This is where you can update your variables
+    if (previousTime_.isNull())
+    {
+        previousTime_ = batteryData.time;
+    }
+    QTime presentTime = batteryData.time;
+    int deltaMseconds;
+    deltaMseconds = abs(presentTime.msecsTo(previousTime_));
```

Comment:
You can combine the above two lines to be `int deltaMseconds = abs(presentTime.msecsTo(previousTime_));`